src/org/rascalmpl/interpreter/utils/JavaFileObjectImpl.java

```diff
@@ -0,0 +1,128 @@
+ * .class output. This uses the
+ * {@link JavaFileObjectImpl#JavaFileObjectImpl(String, JavaFileObject.Kind)}
+ * constructor.
+ * </ol>
+ * This class does not attempt to reuse instances (there does not seem to be a
+ * need, as it would require adding a Map for the purpose, and this would also
+ * prevent garbage collection of class byte code.)
+ */
+ public final class JavaFileObjectImpl extends SimpleJavaFileObject {
+   // If kind == CLASS, this stores byte code from openOutputStream
+   private ByteArrayOutputStream byteCode;
+
+   // if kind == SOURCE, this contains the source text
+   private final CharSequence source;
+
+   private final ISourceLocation sloc;
+
+   /**
+    * Construct a new instance which stores source
+    * 
+    * @param baseName
+    *           the base name
+    * @param source
+    *           the source code
+    */
+   JavaFileObjectImpl(final String baseName, final CharSequence source) throws URISyntaxException {
+      this(null, baseName, source);
+   }
+
+   JavaFileObjectImpl(ISourceLocation sloc, String baseName, final CharSequence source) throws URISyntaxException  {
+      this(sloc, baseName, source, Kind.SOURCE);
+   }
+
+   public ISourceLocation getSloc() {
+      return sloc;
+   }
+   
+   /**
+    * Construct a new instance which stores source
+    * 
+    * @param baseName
+    *           the base name
+    * @param source
+    *           the source code
+    * @throws URISyntaxException 
+    */
+   JavaFileObjectImpl(ISourceLocation sloc, String baseName, final CharSequence source, Kind kind) throws URISyntaxException {
+      super(JavaCompiler.toURI(baseName + JavaCompiler.JAVA_EXTENSION), kind);
+      this.source = source;
+      this.sloc = sloc;
+   }
+
+   /**
+    * Construct a new instance
+    * 
+    * @param name
+    *           the file name
+    * @param kind
+    *           the kind of file
+    */
+   JavaFileObjectImpl(final String name, final Kind kind) throws URISyntaxException {
+      this(null, name, null, kind);
+   }
+
+   /**
+    * Return the source code content
+    * 
+    * @see javax.tools.SimpleJavaFileObject#getCharContent(boolean)
+    */
+   @Override
+   public CharSequence getCharContent(final boolean ignoreEncodingErrors)
+         throws UnsupportedOperationException {
+      if (source == null) {
+         throw new UnsupportedOperationException("getCharContent()");
+      }
+      return source;
+   }
+
+   /**
+    * Return an input stream for reading the byte code
+    * 
+    * @see javax.tools.SimpleJavaFileObject#openInputStream()
+    */
+   @Override
+   public InputStream openInputStream() {
+      return new ByteArrayInputStream(getByteCode());
+   }
+
+   /**
+    * Return an output stream for writing the bytecode
+    * 
+    * @see javax.tools.SimpleJavaFileObject#openOutputStream()
+    */
+   @Override
+   public OutputStream openOutputStream() {
+      byteCode = new ByteArrayOutputStream();
```

Comment:
can it be called more than once? if no, how about we allocate this at ctor time?

Reply:
This is almost never called.

Reply:
Only when we save the generated parser to a jar file.

Reply:
I suspect we have to be able to open and close it multiple times in some cases, but I'm not sure.